Lk reblocking blog (#510)
broadinstitute/warp: website/docs/Pipelines/reblock_gvcf_spec.md

---
sidebar_position: 14
---

> 

Reviewer: Where will this appear in the WARP doc site navigation on the side bar? Perhaps a sub page of the Whole Genome and Exome sections? Or we could group those two under Germline and have it be a top level page there?

# Reblocked GVCF Technical Overview

> The following is a technical overview of the current reblocked GVCF schema (gnomADv4) as well descriptions of the previous GVCF schemas.

Reviewer: ```suggestion
The following is a technical overview of the current reblocked GVCF schema for the Whole Genome and Exome Germline Pipelines (v3.0.0 and newer) as well descriptions of the previous GVCF schemas. Published 11/18/2021.
```

Doesn't have to be this exact wording but I think it's good to note when it's current as of and when it takes effect in the pipelines.

> ## gnomADv4 schema (4 bands)

Reviewer: ```suggestion
## 4 Band Schema (default as of Nov 2021)
```

> * New “current” as of August 2021

Reviewer: ```suggestion
* In use for gnomADv4 and default in Whole Genome and Exome Germline Pipelines v3.0.0 and newer
```

> * Arguments -do-qual-approx --floor-blocks -GQB 20 -GQB 30 -GQB 40 

Reviewer: code styling?

#### Summary (improvements compared with HC GVCFs):

> * NOTE: this format is currently incompatible with GenotypeGVCFs!  Only GnarlyGenotyper can be specified in Joint genotyping WDL input json.

Reviewer: This is no longer true, entire note can be removed

> * PLs are omitted for homozygous reference sites to save space (which is why this format is currently incompatible with GenotypeGVCFs) -- GQs are output for genotypes, PLs can be approximated as [0, GQ, 2*GQ].

Reviewer: also can remove incompatibility () here

* Compared with old HaplotypeCaller GVCFs, GQ resolution for homozygous reference genotypes is reduced -- i.e. homRef GQs will be underconfident, which may affect analyses like de novo calling where confident reference genotypes are important.

* Alleles that aren’t called in the sample genotype are dropped, i.e. each variant should have no more than two alt alleles (in very rare cases, usually just one plus <NON_REF>)
New annotations to enable merging data for filtering without using genotypes -- RAW_GT_COUNT(S?) for doing ExcessHet calculation from a sites-only file; QUALapprox and/or AS_QUALapprox for doing QUAL approximation/filling QUAL VCF field from a combined sites-only field; VarDP and/or AS_VarDP used to calculate QualByDepth/QD annotation for VQSR

> * No more MIN_DP -- joint DP varies?

Reviewer: Should either remove or resolve this question in the text

* InbreedingCoeff values will vary due to some low quality genotypes being rounded down to GQ0

#### Cost/scale improvements

* Storage footprint is reduced compared with HaplotypeCaller output
* Fewer VariantContexts, i.e. lines, per VCF speeds up GenomicsDB/Hail import
* Fewer alternate alleles reduces memory requirements for merging

### Specific improvements compared with 7-band schema
* Not dropping GQ0s -- reblocked GVCF should cover all positions (that input GVCF covers)
* No overlaps -- only overlapping positions should be two variants, i.e. deletions, on separate haplotypes
(No more no-calls in GVCFs-- all genotypes should be called, positions with no data will be hom ref with GQ0; NOTE that this is not being validated and GenotypeGVCFs outputs will have no-calls to preserve previous behavior)

> * Latest Terra WDL and config at https://app.terra.bio/#workspaces/ccdg-gatk-supp-terra/CCDG_WES_Reblocking_Work_F3/workflows/ccdg-gatk-supp-terra/ReblockGVCF-gatk4_exomes_goodCompression_fixed (NOTE that this has an off-by-one issue for contigs starts of exactly 1, which we don’t see in Broad production)

Reviewer: I know the public workspace isn't out yet but maybe we don't want to link to a workspace until it is? I think I'd just remove this entire bullet point, this workspace is outdated.

> 

Reviewer: ```suggestion
Prior to v3.0.0 of the Whole Genome and Exome pipelines, reblocking was not performed by default. It was only performed in preparation for joint calling samples and the pipeline was previously not publicly released. The following sections describe previously used reblocking schemas for various large projects completed at the Broad Institute.
```

## CCDG WGS (7 bands)

> * Suitable for de novo calling from exomes

Reviewer: ```suggestion
* Purpose: Suitable for de novo calling from exomes and used for UKBB
```

> * Arguments -drop-low-quals -do-qual-approx --floor-blocks -GQB 10 -GQB 20 -GQB 30 -GQB 40 -GQB 50 -GQB 60

Reviewer: 'code' format if possible/logical?

> Q* UALapprox output in INFO field, required for GnarlyGenotyper

Reviewer: ```suggestion
* QUALapprox output in INFO field, required for GnarlyGenotyper
```

* Low quality alt alleles will be dropped (below stand-call-conf = 30)
* Alleles that aren’t called will be dropped 
* homRef calls at GQ0 (default) are dropped
* Non-ref AD zeroed out
* GTs with non-ref are converted to GQ0 homRef and dropped
* Blocks are floored -- GQ output for block is the lower bound for that block, regardless of median or min
* Current WDL at https://portal.firecloud.org/?return=terra#methods/methodsDev/ReblockGVCF-gatk4_exomes_goodCompression/4

> * As used for UKBB

Reviewer: delete line, moved above

### gnomADv3 schema (deprecated -- 2 bands)

> Arguments -drop-low-quals -do-qual-approx -rgq-threshold 10

Reviewer: can arguments be put into 'code' format?

* Default GQ blocks, < 20 and >= 20
* homRef calls below GQ10 are dropped
* → remaining tranche of homRef calls between GQ 10 and 20 gnomAD team had to filter
* QUALapprox output in INFO field, required for GnarlyGenotyper
* Low quality alt alleles will be dropped (below stand-call-conf = 30)
* Extra INFO annotations dropped, only data required for VQSR retained: DP, MQ, ReadPosRankSum, MQRankSum (strand bias data is in FORMAT)
* Non-ref AD zeroed out

> * WDL at https://portal.firecloud.org/?return=terra#methods/eric-methods/ReblockGVCF-gatk4/1/wdl

Reviewer: made into a link?

> ## Original HaplotypeCaller schema (current Warp single-sample pipeline)

Reviewer: no longer 'current' - WARP single sample prior to v3.0.0, right?

Reviewer: Also maybe worth noting that this is without reblocking, this is the standard output from haplotype caller?
